### PR TITLE
add a first draft of the discovery lib

### PR DESCRIPTION
**Pull request by natebrennand**
cc @kvigen @mohit

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,2 +1,15 @@
...
+## Usage
+
+#### func  Discover
+
+```go
+func Discover(service, name string) string
+```
+Discover finds the specified URL for a service based off of the service's name
+and which interface you are accessing. Values are found in environment variables
+fitting the scheme: SERVICE_{SERVICE NAME}_{INTERFACE NAME}_{PROTO,HOST,PORT}.
```
Comment (rgarcia) [inline README.md]: If this is public you can make your life easier and link to the godoc, right?

Comment (natebrennand) [inline README.md]: amended. won't work until it merges though

File: configure.go [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+
+const (
+	templateVar = "SERVICE_%s_%s_%%s"
+)
+
+func getVar(envVar string) string {
+	envVar = strings.ToUpper(envVar)
+	val := os.Getenv(envVar)
+	if val == "" {
+		panic(kv.FormatLog("discovery-go", kv.Error, "missing env var", m{
```
Comment (dfuentes) [inline configure.go]: Libraries should not panic.

File: configure.go [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+
+// m is a convenience type for using kv.
+type m map[string]interface{}
+
+const (
+	templateVar = "SERVICE_%s_%s_%%s"
+)
+
+func getVar(envVar string) string {
+	envVar = strings.ToUpper(envVar)
```
Comment (rgarcia) [inline configure.go:21]: nit: env variable names are case sensitive, I think? We don't use any lowercase env variable names so this is probably fine

Comment (natebrennand) [inline configure.go:21]: Yes, they are case sensitive. This is so you can pass in a service's name lowercase.

File: configure.go (hunk 2 above)
Comment (natebrennand) [inline configure.go]: Environment variables should always be present when using this library. Do you think it is better to return an error?

File: configure.go [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+			"var": envVar,
+		}))
+	}
+	return val
+}
+
+// Discover finds the specified URL for a service based off of the service's name and which
+// interface you are accessing. Values are found in environment variables fitting the scheme:
+// SERVICE_{SERVICE NAME}_{INTERFACE NAME}_{PROTO,HOST,PORT}.
+func Discover(service, name string) string {
```
Comment (dfuentes) [inline configure.go]: I would rename this to better indicate it is returning a url.  I would also add functions for getting the host/port/proto on their own.

File: configure.go (hunk 2 above)
Comment (dfuentes) [inline configure.go]: Definitely.

File: configure.go (hunk 4 above)
Comment (natebrennand) [inline configure.go]: amended to `DiscoverURL`

**Comment by dfuentes** [comment]
Not to harp on this too much, but I would think through the naming a bit more.  It seems redundant to have Discover in all the names (they will be called as `discovery.DiscoverX`).  Maybe it would be better as `discovery.X`?

Feel free to ignore, rest looks fine.

File: configure.go [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+			"var": envVar,
+		}))
+	}
+	return val, nil
+}
+
+// URL finds the specified URL for a service based off of the service's name and which
+// interface you are accessing. Values are found in environment variables fitting the scheme:
+// SERVICE_{SERVICE NAME}_{INTERFACE NAME}_{PROTO,HOST,PORT}.
+func URL(service, name string) (string, error) {
```
Comment (kvigen) [inline configure.go:34]: sorry for being late to this PR, but i would curious what the INTERFACE_NAME is for. Is SERVICE_NAME not enough?

Comment (natebrennand) [inline configure.go:34]: Better late than never!

Adding INTERFACE_NAME accounts for a handful of services that expose multiple ports with different protocols. [Influxdb-service](https://github.com/Clever/influxdb-service/blob/master/launch/influxdb-service.yml) is one example.
There aren't a ton of services that have multiple interfaces but this gives us a lot of future extendability.

Comment (kvigen) [inline configure.go:34]: Sounds reasonable. Thanks for the explanation

**Pull request merged**